README.md

```diff
@@ -1,81 +1,149 @@
 
-Multiaddr is a standard way to represent addresses that: 
-- Support any standard network protocols.
-- Self-describe (include protocols).
-- Have a binary packed format.
-- Have a nice string representation.
-- Encapsulate well.
 
 ## Specification
 
-Normally, addresses have been represented using string addresses, like:
+- Human-readable multiaddr: `(/<protoName string>/<value string>)+`
+  - Example: `/ip4/1.2.3.4/tcp/80/http/example.com`
+- Machine-readable multiaddr: `(<protoCode uvarint><value []byte>)+`
+  - Same example: `0x4 0x01020304 0x6 0x3830 0x480 0x6578616d706c652e636f6d` (TODO: doublecheck the bytes)
+  - Values are usually length-prefixed with a uvarint
+
+Multiaddr and all other multiformats use unsigned varints (uvarint).
+Read more about it in [multiformats/unsigned-varint](https://github.com/multiformats/unsigned-varint).
+
 
-```
-tcp4://127.0.0.1:1234
-udp4://10.20.30.40:5060
-ws://1.2.3.4:5678
-tcp6://[1fff:0:a88:85a3::ac1f]:8001
-```
+### Encoding
 
-This isn't optimal. Instead, addresses should be formatted so:
+TODO: specify the encoding (byte-array to string) procedure
 
-### Binary format:
+### Decoding
 
-```
-(<varint proto><n byte addr>)+
-<1 byte ipv4 code><4 byte ipv4 addr><1 byte udp code><2 byte udp port>
-<1 byte ipv6 code><16 byte ipv6 addr><1 byte tcp code><2 byte tcp port>
-```
+TODO: specify the decoding (string to byte-array) procedure
 
-### String format:
 
-```
-(/<addr str code>/<addr str rep>)+
-/ip4/<ipv4 str addr>/udp/<udp int port>
-/ip6/<ipv6 str addr>/tcp/<tcp int port>
-```
+## Protocols
 
-### Protocols
+See [protocols.csv](protocols.csv) for a list of protocol codes and names,
+and [protocols/](protocols/) for specifications of the currently supported protocols.
 
-See [protocols.csv](protocols.csv).
+TODO: most of these are way underspecified
+
+- /ip4, /ip6
+- /dns4, /dns6
+- /dnsaddr
+- /tcp
+- /udp
+- /utp
+- /tls
+- /ws, /wss
+- /ipfs
+- /p2p-circuit
+- /p2p-webrtc-star, /p2p-webrtc-direct
+- /p2p-websocket-star
+- /onion
 
-Originally from here:
-https://github.com/jbenet/random-ideas/issues/11
 
 ## Implementations
 
 - [js-multiaddr](https://github.com/multiformats/js-multiaddr) - stable
 - [go-multiaddr](https://github.com/multiformats/go-multiaddr) - stable
+  - [go-multiaddr-dns](https://github.com/multiformats/go-multiaddr-dns)
+  - [go-multiaddr-net](https://github.com/multiformats/go-multiaddr-net)
+  - [go-multiaddr-filter](https://github.com/multiformats/go-multiaddr-filter)
+  - [go-multiaddr-fmt](https://github.com/multiformats/go-multiaddr-fmt)
 - [java-multiaddr](https://github.com/multiformats/java-multiaddr) - stable
 - [haskell-multiaddr](https://github.com/MatrixAI/haskell-multiaddr) - stable
 - [py-multiaddr](https://github.com/sbuss/py-multiaddr) - alpha
 - [rust-multiaddr](https://github.com/multiformats/rust-multiaddr) - beta
 - [cs-multiaddress](https://github.com/tabrath/cs-multiaddress) - alpha
```

Comment:
Should be /multiformats/cs-multiaddress as it's moved to the org, also remove the alpha tag; it's pretty stable now.